manual/core/ssl/README.md

```diff
@@ -0,0 +1,159 @@
+pair for the client:
+
+```
+keytool -genkey -keyalg RSA -alias client -keystore client.keystore
+```
+
+If you're using self-signed certificates, extract the public part of the client certificate, and
+import it in the truststore of each Cassandra node:
+
+```
+keytool -export -alias client -file client.cer -keystore client.keystore
+keytool -import -v -trustcacerts -alias client -file client.cer -keystore server.truststore
+```
+
+If you're using a CA, sign the client certificate with it (see the blog post linked at the top of
+this page). Then the nodes' truststores only need to contain the CA's certificate (which should
+already be the case if you've followed the steps for inter-node encryption).
+
+
+### Driver configuration
+
+By default, the driver's SSL support is based on the JDK's built-in implementation: JSSE (Java
+Secure Socket Extension),.
+
+To enable it, you need to define an engine factory in the [configuration](../configuration/).
+
+#### JSSE, property-based
+
+```
+datastax-java-driver {
+  advanced.ssl-engine-factory {
+    class = DefaultSslEngineFactory
+    
+    # This property is optional. If it is not present, the driver won't explicitly enable cipher
+    # suites on the engine, which according to the JDK documentations results in "a minimum quality
+    # of service".
+    // cipher-suites = [ "TLS_RSA_WITH_AES_128_CBC_SHA", "TLS_RSA_WITH_AES_256_CBC_SHA" ]
+  }
+}
+```
+
+You can then use [JSSE system properties] for specific details, like keystore locations and
+passwords:
+
+```
+-Djavax.net.ssl.trustStore=/path/to/client.truststore
+-Djavax.net.ssl.trustStorePassword=password123
+# If you're using client authentication:
+-Djavax.net.ssl.keyStore=/path/to/client.keystore
+-Djavax.net.ssl.keyStorePassword=password123
+```
+
+#### JSSE, programmatic
+
+If you need more control than what system properties allow, you need to write your own engine
+factory. If you just need specific configuration on the `SSLEngine`, you can extend the default
+factory and override `newSslEngine`. For example, here is how to enable hostname verification:
+
+```java
+public class HostnameVerificationSslEngineFactory extends DefaultSslEngineFactory {
+
+  public HostnameVerificationSslEngineFactory(DriverContext context) {
+    super(context);
+  }
+
+  @Override
+  public SSLEngine newSslEngine(SocketAddress remoteEndpoint) {
+    SSLEngine engine = super.newSslEngine(remoteEndpoint);
+    SSLParameters parameters = engine.getSSLParameters();
+    // HTTPS endpoint identification includes hostname verification against certificate's common
+    // name.
+    parameters.setEndpointIdentificationAlgorithm("HTTPS");
+    engine.setSSLParameters(parameters);
+    return engine;
+  }
+}
+```
+
+Then declare your custom implementation in the configuration:
+
+```
+datastax-java-driver {
+  advanced.ssl-engine-factory {
+    class = com.mycompany.HostnameVerificationSslEngineFactory
+  }
+}
+```
+
+#### Netty
```

Comment:
I wonder if we should consider providing something for this in the form of a separate module, what do you think?  I think providing it in the core module was problematic for us in 3.x, but adding a separate module for it might be a nice compromise as the Netty SSL implementation is much more object allocation efficient / performant.

Reply:
Created: https://datastax-oss.atlassian.net/browse/JAVA-1859